File: Darts/src/code/bean/GameParamFilterPanelX01.java
package code.bean;
import java.awt.BorderLayout;
import java.awt.event.ActionListener;

import javax.swing.JPanel;

public final class GameParamFilterPanelX01 extends GameParamFilterPanel
{
	public GameParamFilterPanelX01()
	{
		super();
		
		add(panel, BorderLayout.CENTER);
		panel.add(spinner);
	}
	
	private final JPanel panel = new JPanel();
	private final SpinnerX01 spinner = new SpinnerX01();

	@Override
	public String getGameParams()
	{
		return "" + spinner.getValue();
	}
	@Override
	public void setGameParams(String gameParams)
	{
		int x01 = Integer.parseInt(gameParams);
		spinner.setValue(x01);
	}
	
	@Override
	public String getFilterDesc() 
	{
		return "games of " + getGameParams();
	}

	@Override
	public void enableChildren(boolean enabled)
	{
		spinner.setEnabled(enabled);
	}
	
	@Override
	public void addActionListener(ActionListener listener)
	{
		spinner.addActionListener(listener);
	}
	
}

File: src/main/kotlin/dartzee/bean/GameParamFilterPanelX01.kt
package dartzee.bean

import dartzee.game.FinishType
import dartzee.game.X01Config
import java.awt.BorderLayout
import java.awt.event.ActionListener
import javax.swing.Box
import javax.swing.JCheckBox
import javax.swing.JPanel

class GameParamFilterPanelX01 : GameParamFilterPanel() {
    private val panel = JPanel()
    private val spinner = SpinnerX01()
    private val cbFinishOnDouble = JCheckBox("Finish on double")
    private val separator = Box.createHorizontalStrut(20)

    init {
        add(panel, BorderLayout.CENTER)
        panel.add(cbFinishOnDouble)
        panel.add(separator)
        panel.add(spinner)
        cbFinishOnDouble.isSelected = true
    }

    override fun getGameParams() = constructGameParams().toJson()

    override fun setGameParams(gameParams: String) {
        val config = X01Config.fromJson(gameParams)
        spinner.value = config.target
        cbFinishOnDouble.isSelected = config.finishType == FinishType.Doubles
    }

    private fun constructGameParams() =
        X01Config(
            spinner.value as Int,
            if (cbFinishOnDouble.isSelected) FinishType.Doubles else FinishType.Any
        )

    override fun getFilterDesc() = "games of ${constructGameParams().description()}"

    override fun enableChildren(enabled: Boolean) {
        spinner.isEnabled = enabled
        cbFinishOnDouble.isEnabled = enabled
    }

    override fun addActionListener(listener: ActionListener) {
        spinner.addActionListener(listener)
        cbFinishOnDouble.addActionListener(listener)
    }

    override fun removeActionListener(listener: ActionListener) {
        spinner.removeActionListener()
        cbFinishOnDouble.addActionListener(listener)
    }
}
